### PR TITLE
Remove `AcceleratorConnector.tpu_cores` and Deprecate `Trainer.tpu_cores`

### PR DESCRIPTION
**Pull request by DuYicong515**
## What does this PR do?

Part of #11449 : remove unused properties in accelerator_connector: tpu_cores
Part of #12171 : deprecate device related properties in Trainer that is accessed from accelerator_connector: tpu_cores

`AcceleratorConnector.tpu_cores` was not meant to be public. It's removed from AcceleratorConnector and deprecated in Trainer in favor of using `Trainer.num_devices`

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/callbacks/xla_stats_monitor.py [hunk 1]
```diff
@@ -75,8 +75,8 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
 
         if isinstance(trainer.accelerator, TPUAccelerator):
             raise MisconfigurationException(
-                "You are using XLAStatsMonitor but are not running on TPU"
-                f" since `tpu_cores` attribute in Trainer is set to {trainer.tpu_cores}."
+                "You are using XLAStatsMonitor but are not running on TPU."
+                f" The accelerator type is set to {trainer.accelerator.name().upper()}."
```
Comment (kaushikb11) [inline pytorch_lightning/callbacks/xla_stats_monitor.py:79]: `accelerator.name()` this doesn't exist anyone Ref: #12180

Comment (DuYicong515) [reply pytorch_lightning/callbacks/xla_stats_monitor.py:79]: Good catch!

BTW I think L76 should be `if not isinstance(trainer.accelerator, TPUAccelerator)` instead, do you know why pytorch_lightning/callbacks/xla_stats_monitor.py was not included in TPU tests?

File: pytorch_lightning/callbacks/xla_stats_monitor.py [hunk 2]
```diff
@@ -73,10 +73,10 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
         if not trainer.loggers:
             raise MisconfigurationException("Cannot use XLAStatsMonitor callback with Trainer that has no logger.")
 
-        if isinstance(trainer.accelerator, TPUAccelerator):
+        if not isinstance(trainer.accelerator, TPUAccelerator):
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/xla_stats_monitor.py:76]: This change is not reflected in the tests I believe. Can you double check that the error message gets correctly caught/asserted? Thanks

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/callbacks/xla_stats_monitor.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/xla_stats_monitor.py:76]: looks like there is no test for it.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2100,7 +2102,11 @@ def root_gpu(self) -> Optional[int]:
 
     @property
     def tpu_cores(self) -> int:
-        return self._accelerator_connector.tpu_cores
+        rank_zero_deprecation(
+            "`Trainer.tpu_cores` is deprecated in v1.6 and will be removed in v1.8. "
+            "Please use `Trainer.devices` instead."
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2107]: is it `devices` or `num_devices`? since in changelog it's `num_devices`.

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/callbacks/xla_stats_monitor.py (hunk 2 above)
Comment (DuYicong515) [reply pytorch_lightning/callbacks/xla_stats_monitor.py:76]: I added a new test case in `tests/deprecated_api/test_remove_1-7.py::test_xla_stats_monitor_tpu_not_used` to validate it.

File: tests/deprecated_api/test_remove_1-7.py [hunk 4]
```diff
@@ -516,3 +518,17 @@ def post_dispatch(self, trainer):
 
     with pytest.deprecated_call(match=escape("`CustomPlugin.post_dispatch()` has been deprecated in v1.6")):
         CustomPlugin(torch.device("cpu"))
+
+
+def test_xla_stats_monitor_tpu_not_used(monkeypatch):
+    monkeypatch.setattr(pytorch_lightning.callbacks.xla_stats_monitor, "_TPU_AVAILABLE", lambda: True)
```
Comment (awaelchli) [inline tests/deprecated_api/test_remove_1-7.py:524]:
```suggestion
    monkeypatch.setattr(pytorch_lightning.callbacks.xla_stats_monitor, "_TPU_AVAILABLE", True)
```

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/callbacks/xla_stats_monitor.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/xla_stats_monitor.py]: It will be resolved by #11098

**Pull request merged**